Fix DataObject initialization

vladsud opened the pull request:
The scenarios are described in code in form of comments.
as well as in issue #3628: DataObject initialization call is not guaranteed to complete before create promise is resolved

But the key broken things are first two items from this list, the rest is unrelated to original problems, but also addressed:
1. We do not create / fully initialize DataObject on creation of new data store, resulting in empty snapshot (no DDSs) created if it's gets attached right away
2. There is no way for DataObject to modify behaviors of data store runtime, as DataObject is created only on use (which only happens if some request is processed by data store that resolves to DataObject). This is hard to fix fully due (see below).
3. \<P> generic argument is misleading - it represents optional providers, and other places use \<O> for it. Changing P -> O.
4. Adding preInitialize() hook as 0.26 integration shows that folks are tapping into internal methods to do pre-initialization (at least that's my reading of the code)
5. I experimented with mixins as a way to extend, replace or adjust FluidDataStoreRuntime behaviors.
I like it more than providing hooks for each specific case (like request handler) - it looks rather neat, generic and flexible. In future, we might want to remove registerRequestHandler() in favor of mixins.

This implementation is a first step on the way to address this area - it does not address all issues of # 2. It's impossible to fully address it as it would create deadlock - DDS handles write out full paths from root (container runtime), so resolving them (while loading DataStore) results in deadlock. We need to change that and add back compat code to do the same for existing files at rest that are already using full paths. This work needs to be done in parallel.

vladsud committed:
An attempt to incrementally fix DataObject initialization

vladsud committed:
Detached creation & UTs

vladsud committed:
Change last Edited factory (mostly for future) to not delay create objects

vladsud committed:
Renamed P -> O generic type, as it rerpresents "Optional", and P is very confusing.
Extracting all init logic into proxy class.
Experimenting with mixins as extensibility mechanism for ContainerRuntime.

vladsud commented on packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts:
```diff
@@ -101,26 +106,135 @@ class FluidDataObjectFactory {
+        // Please note that ideal solution is for code above to await instanceP, but given that we can't
+        // do it yet for existing data store, we have to do it only when summary is called, i.e. after
+        // data store runtime is fully initialized (from POV of container runtime)
+        if (!onDemandInstantiation) {
```
Note: I'm thinking about moving this out from here in next iterations by plumbing through default argument:
```
runtimeCtor: typopeof FluidDataStoreRuntime = FluidDataStoreRuntime
```
That way DataObject can provide a bunch of custom behaviors, without complicating interfaces / implementaitons.

vladsud replied:
Fixed

vladsud commented on packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts:
```diff
@@ -101,26 +106,135 @@ class FluidDataObjectFactory {
+        // becomes globally available. But it's not full initialization - constructor can't
+        // access DDSs or other services of runtime as objects are not fully initialized.
+        // In order to use object, we need to go through full initialization by calling getInstance().
+        if (!onDemandInstantiation) {
```
Ideally, this we always create DataObject, but not initialize it until needed (and remove onDemandInstantiation).
I.e. all expensive work happens  in initializingFromExisting / initializingFirstTime / hasInitialized methods, but ctor is cheap.
@leeviana , @aboktor, do you think this is reasonable to change, or it's part of life?
AtMentionsCollection seems like the biggest offender here. All other examples seems to have trivial ctors.

leeviana replied:
Yeah, I think constructor is generally pretty cheap, although there are at least 4 places in component lifecycle that I'm aware of that expensive work can happen and it's not just initialize:

- in the factory create (for getting services, setting them up etc. because you have access to context/runtime, although I think now most of this can happen in initialize), this includes constructor cost
- in initialize
- in request (this is important/key for the scriptor path because until request happens, scriptor does not know extra context things like maybe start/end marker. Thus scriptor can't initialize it's model (with virtualization etc.) until the request
- in render (just because render is expensive)

vladsud replied:
So, do I hear that I should move in this direction and fix AtMentionsCollecitons' ctor to be cheap?

leeviana replied:
Yes, AtMentionsCollections ctor logic can most likely move to initialize :)

vladsud replied:
Switching to always creating DataObjects, but finish initialization only on demand
I'll raise PR in parallel for At-mention collection.

vladsud committed:
cleanup

vladsud commented:
@skylerjokiel , @ChumpChief , @heliocliu - I'd appreciate if you could take a look. Thanks!

vladsud commented on packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts:
```diff
@@ -163,32 +163,60 @@ export class PureDataObjectFactory<TObj extends PureDataObject<P, S>, P, S>
+        // Please note that ideal solution is for code above to await instanceP, but given that we can't
+        // do it yet for existing data store, we have to do it only when summary is called, i.e. after
+        // data store runtime is fully initialized (from POV of container runtime)
+        if (!this.onDemandInstantiation) {
```
I'm not sure what you mean. Nobody today passes onDemandInstantiation = false, so this code does not impact any excising users. But for implementation of search flow, we need components like AtMentionsColleciton to be loaded by summarizer and have a chance to install hooks into runtime to be called and provide extra summary blobs. This code is doing just that - giving opportunity to boot DataObject as part of booting Data store runtime for them to interact.
Also take a look at latest iteration and comment there - I want to move this logic out of DataObject layer by allowing data objects (and their factories) to supply identify of data store runtime (i.e. be able to pass in derived class) instead of having these kinds of custom solutions in here

heliocliu commented on packages/framework/aqueduct/src/data-object-factories/dataObjectFactory.ts:
```diff
@@ -24,16 +24,16 @@ import { PureDataObjectFactory } from "./pureDataObjectFactory";
+ * O - represents a type that will define optional providers that will be injected
  * S - the initial state type that the produced data object may take during creation
  */
-export class DataObjectFactory<TObj extends DataObject<P, S>, P, S> extends PureDataObjectFactory<TObj, P, S>
```
fwiw it looks like the `O` usage elsewhere is to distinguish between optional and required providers, whereas here there's no distinction which is why this was `P` instead.  imo `P` would still make more sense here but not a big deal either way

vladsud replied:
Can you please elaborate? The comment above says P - is for optional providers.
Digging deeper, it maps to 
```
protected readonly providers: AsyncFluidObjectProvider<FluidObjectKey<P>, FluidObjectKey<object>>;

export type AsyncFluidObjectProvider<O extends keyof IFluidObject, R extends keyof IFluidObject>
    = AsyncOptionalFluidObjectProvider<O> & AsyncRequiredFluidObjectProvider<R>;
```
as well as 
```
    const providers = dependencyContainer.synthesize<P>(this.optionalProviders, {});

    public synthesize<
        O extends IFluidObject,
        R extends IFluidObject = {}>(
            optionalTypes: FluidObjectSymbolProvider<O>,
            requiredTypes: FluidObjectSymbolProvider<R>,
    ): AsyncFluidObjectProvider<FluidObjectKey<O>, FluidObjectKey<R>> {

```
So it's consistent through code base P stands for optional providers.
Did I miss something?

heliocliu replied:
didn't miss anything, i uh... just didn't really have anything substantial to say i guess 😐

heliocliu commented on packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts:
```diff
@@ -101,26 +106,135 @@ class FluidDataObjectFactory {
+        readonly context: IFluidDataStoreContext,
+        private readonly sharedObjectRegistry: ISharedObjectRegistry,
+        private readonly optionalProviders: FluidObjectSymbolProvider<O>,
+        private readonly props: S,
```
For the use cases does it make sense to provide props as an arg through `getInstance` instead?  That way the data object isn't holding onto this forever after using it

vladsud replied:
getInstance() can be called from within object - it's actually main path, so it has to have props in there.

heliocliu commented on packages/framework/aqueduct/src/data-objects/pureDataObject.ts:
```diff
@@ -190,6 +191,12 @@ export abstract class PureDataObject<P extends IFluidObject = object, S = undefi
+     * Called every time the data store is initialized, before initializingFirstTime or
+     * initializingFromExisting is called.
+     */
+    protected async preInitialize(): Promise<void> { }
```
thoughts on something like `willInitialize` instead? matches `hasInitialize`/the other method names better i think

vladsud replied:
I feel like hasInitialize should be changed to hasInitialized :)
What do you think?

heliocliu replied:
Whoops, that's a typo on my side; it's already `hasinitialized`

vladsud committed:
Merge branch 'main' of https://github.com/Microsoft/FluidFramework into DataObjecrInit

# Conflicts:
#	packages/framework/aqueduct/src/data-objects/dataObject.ts
#	packages/framework/aqueduct/src/data-objects/pureDataObject.ts

vladsud committed:
Игшдв икуфл

vladsud committed:
Remove summary mixin
remove on-demand creation (DataObjects are created always)
Add ability to provide your own runtie ctor

vladsud committed:
Braking.md

vladsud commented:
@leeviana, can you please glance at breaking.md to make sure it's clear what is the impact and you are Ok with that,.

leeviana commented:
Yes, the breaking.md seems clear, I think it just brings the lifecycle methods closer to what we expected to happen anyways. After this change at least the initialize method will be in the initial summary ops, although we still might have to spend some more time thinking about how to allow callers to "do initialization work" without initial config, which may require other lifecycle methods being called that set things that we would want in the initial op. (i.e. initial # of rows in a table)
